import "kernel/mango/proto/common.proto";       // for NMango::TIndexQuotes
import "yweb/protos/hostfactors.proto";
import "yweb/protos/docfactors.proto";
import "yweb/protos/orange.proto";
import "yweb/protos/spamchecker.proto";

option java_package = "NRobotRoot";

message TLinkPreparatData {
    optional string Url = 1;
    optional string Text = 2;
    optional uint32 Flags = 3;
    optional uint32 SeoFlags = 4;
    optional uint32 BanFlags = 5;
    optional uint32 Language = 6;
    optional uint32 SourceDate = 7;
    optional uint32 LinkDate = 8;
    optional double SourceRank = 9;
}

// %% {{a name="TAnchorText"}} **~Anchor Text collection** %%
message TAnchorText {
    message TExtraHostFactors {
        optional uint32 OwnerTic       = 1; // TIC of host owner
        optional uint32 OwnerGeo       = 2; // Geo of host owner
        repeated uint32 OwnerCatty     = 3; // Catty of host owner
    }

    message TAnchorTextEntry {
        optional string Url                         = 1; // Source URL
        optional string AnchorText                  = 2; // anchor text
        optional fixed32 Flags                      = 3; // TBD
        optional double CrawlRank                   = 4; // Crawlrank of the link
        optional fixed32 DiscoveryTime              = 5; // Time when link was found
        optional string Language                    = 6; // Language of source URL
        optional TExtraHostFactors ExtraHostFactors = 7;
    }

    message TAnchorTextStats {
        optional uint32 InternalReferrers = 1;
        optional uint32 ExternalReferrers = 2;
        optional uint32 UniqueAnchorTexts = 3;
        optional uint32 UniqueAnchorOwners = 4;
    }

    repeated TAnchorTextEntry AnchorText            = 1; // anchor text
    optional double CrawlRank                       = 2; // Crawlrank of the doc
    optional uint32 AddTime                         = 3; // Orange DocInfo creation time or time of first link discovery
    optional int32  CrawlRankErf                    = 4; // Value to be stored in ERF, must belong to [0..255]
    optional double MaxHostCrawlRank                = 6; // Max Crawkrank of the host, was not calculated if not set
    optional NRealTime.THostFactors HostFactors     = 9;
    optional TExtraHostFactors ExtraHostFactors     = 10;
// Don't reuse 'DocFactors' id with another type till 2013/03/21
    optional NRealTime.TDocFactors DocFactors       = 11;
    optional fixed32 HostAddTime                    = 12; // how long orange knows about the host
    optional uint32  HostSize                       = 13; // number of "good" docs on host
    optional TAnchorTextStats AnchorStats           = 15;
    optional fixed32 ExportTime                     = 16; // Time when exported from Orange
    optional fixed32 PubTime                        = 17; // Publication time from RSS
    optional string HostLanguage                    = 18; // host language
// Don't reuse 'LinkBadness' id with another type till 2013/03/21
//  optional uint32 LinkBadness                     = 19; // 0..255
    optional double PornoWeight                     = 22; // Deprecated, use SpamFactors

    repeated NOrangeData.TDocAnnotation Annotations = 7;
    repeated NOrangeData.TDocListInfo DocListInfo   = 20; // doclists the doc belongs to
    repeated NOrangeData.TExportRecord ExportStream = 21;
    repeated NOrangeData.TAttribute SearchAttribute = 23; // Attributes to be added into search database

    optional NRealTime.TSpamFactors SpamFactors     = 24; // Factors calculated using antispam rules

    // Comment for services: if you have "abused" this logic and created enormously long-living document,
    // you can remove it by a duplicate with shorter live time or by UpdUrl anti-document.
    optional uint64 RemoveTime                      = 25; // Remove time need for rtsearch
    optional NMango.TIndexQuotes MangoQuotes        = 26; // Mango data
    repeated NOrangeData.TTimeAnnotation TimeAnnotations = 27; // timing datas
}

// %% {{a name="TFeedWithContent"}} **~FeedWithContent %%
message TFeedWithContent {
        required bytes Logel = 1;       // content from external feed source
        optional string Annotation = 2; // optional meta data
        optional uint32 Rating = 3; //Rating (column 3 in fastcontent files)
        optional uint32 Readers = 4; //Readers count (column 4 in fastcontent files)
}


// Deprecated message, left for compatibility with existing Site Search code only.
///Дополнительная информация по урлу в быстророботной базе
message TUrlExtraData {
    message TSourceInfo {
        // according to config in https://arcadia.yandex.ru/arc/trunk/arcadia/yweb/common/fastconf/sources-fast.cfg
        required uint32 SID                 = 1 [ default = 0   ] ; // Source ID, 2'nd column of sources-fast.cfg
        optional uint32 Hops                = 2 [ default = 255 ]; // Hops: (0,1,2)
        optional fixed32 LastAddTime        = 3; // When Url was added to doclist last time
    }

    repeated TSourceInfo SourceInfo              = 1;
    optional uint32 LeastAddTime                 = 2;
    optional string DocLists                     = 3;
    optional uint32 AnnAddTime                   = 4;
}

message TSourceCountry {
    required uint32 Id          = 1;
    required fixed32 TimeStamp  = 2;
}

message TUrlExtData {
    optional uint32 LastAccessTrue          = 1;  // to keep it in cluster url dat
    optional double CrawlRank               = 2;  // for merging info from orange in updurl
    optional double CrawlRankExt            = 3;  // for merging info from orange in updurl
    optional double ForeignDMOZPageRank     = 4;
    optional uint32 LastTimeExistLink       = 5;
    optional string MultilanguageHostPrefix = 6;
    optional NRealTime.TDocFactors.TDocListsFactors DocListsFactors = 7; // Deprecated, see DocListInfo
    optional fixed32 LastExportTime         = 8;
    optional uint32 LatestTimeDocIdChanged  = 9;
    optional int32 LastModifiedTimeKnownFromSitemap = 10; //could be used only in policies based on sitemap info!
    optional bool HasStructuredData         = 11;
    repeated NOrangeData.TDocListInfo DocListInfo     = 12;
    repeated NOrangeData.TExportRecord ExportStream   = 13;
    optional NRealTime.TDocFactors.TLinkStat LinkStat = 14;
    optional double ForeignTRPageRank       = 15;
    optional fixed64 IncomingRssLinkDate    = 16;
    optional bool IsFromSitemap             = 17;
    optional uint32 ChangeFrequencyBySitemapInfo = 18; // in seconds
    optional double PriorityBySitemapInfo = 19;
    optional uint32 SourceId = 20;
    // UNUSED 21
    optional uint32 FirstAccess = 22;
    optional uint64 StatusHistory = 23;
    optional bool ForceCrawl = 24;
    optional uint32 DateOfForceCrawlSetting = 25;
    repeated TSourceCountry SourceCountry = 26;
    optional bool IsFakeForRedirect = 27;
    optional bool IsFakeForBan = 28;
    optional bool LangRegion = 29;
}

message TPagerankResults {
    repeated float  Pr     = 1;
    optional float  Dmoz   = 2;
    optional float  Turkey = 3;
    optional uint64 Edges  = 4;
}

message TUrlExtCrawlData {
    optional double CrawlRank           = 1;
    optional double DmozPR              = 2;
    optional double TurkeyPR            = 3;
    optional double RussianRR           = 4;
    optional double ForeignRR           = 5;
    optional double RussianSR           = 6;
    optional double ForeignSR           = 7;
    optional double TurkeySR            = 8;
    optional double UkrainianSR         = 9;
    optional double RussianPR           = 10;
    optional double UkrainianPR         = 11;
    optional double DeadRank            = 12;
    optional double Rank404             = 13;
    optional double Variability         = 14;
    optional uint32 UpdateInterval      = 15;
    optional uint32 MaxInterval         = 16;
    optional uint32 ShingleSize         = 17;
    optional bool IsOld                 = 18;
}

message TRssLinkExtData {
    optional string Description = 1;
    optional string Date        = 2;
    optional string Anchor      = 3;
    optional string Author      = 4;
    optional string Source      = 5;
}
